rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
+	}()
 
-		// Ensure at least one heartbeat succeeds before exiting the
-		// retry loop. If it fails, don't retry: The node is probably
-		// dead.
-		if err = c.heartbeat(); err != nil {
-			return err
-		}
+	return newCall
+}
 
-		return nil
+func (c *Client) connect() error {
+	conn, err := tlsDialHTTP(c.network, c.remoteAddress, c.tlsConfig)
+	if err != nil {
+		return err
 	}
+	c.localAddress = conn.LocalAddr().String()
+	c.rpcClient = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
 
-	return util.Errorf("system is stopping")
+	return nil
 }
 
-// IsConnected returns whether the client is connected.
-func (c *Client) IsConnected() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.Client != nil
+// Healthy returns a channel that is closed when the client becomes healthy.
+func (c *Client) Healthy() <-chan struct{} {
+	return c.healthy.Load().(chan struct{})
 }
 
-// IsHealthy returns whether the client is healthy.
-func (c *Client) IsHealthy() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.healthy
-}
-
-// Addr returns remote address of the client.
-func (c *Client) Addr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.addr
-}
+func (c *Client) setHealthy() {
+	if c.isHealthy {
+		return
+	}
 
-// LocalAddr returns the local address of the client.
-func (c *Client) LocalAddr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.lAddr
+	c.isHealthy = true
+	close(c.healthy.Load().(chan struct{}))
 }
 
-// RemoteOffset returns the most recently measured offset of the client clock
-// from the remote server clock.
-func (c *Client) RemoteOffset() proto.RemoteOffset {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.offset
+func (c *Client) setUnhealthy() {
+	if c.isHealthy {
+		c.isHealthy = false
+		c.healthy.Store(make(chan struct{}))
+	}
 }
 
-// Close removes the client from the clients map and closes
-// the Closed channel.
+// Close closes the Closed channel, which triggers the end of the run loop and
+// removal from the clients map.
 func (c *Client) Close() {
 	clientMu.Lock()
-	if !c.closed {
-		if c.cached {
-			delete(clients, c.Addr().String())
-		}
-		c.mu.Lock()
-		c.healthy = false
-		c.closed = true
-		if c.Client != nil {
-			c.Client.Close()
-		}
-		c.mu.Unlock()
-		close(c.Closed)
+	defer clientMu.Unlock()
+
+	if c.isClosed {
+		return
 	}
-	clientMu.Unlock()
+
+	c.isClosed = true
+	delete(clients, c.key)
+	close(c.Closed)
 }
 
 // runHeartbeat sends periodic heartbeats to client. Closes the
 // connection on error. Heartbeats are sent in an infinite loop until
 // an error is encountered.
-func (c *Client) runHeartbeat(stopper *stop.Stopper) {
-	if log.V(2) {
-		log.Infof("client %s starting heartbeat", c.Addr())
-	}
-
+func (c *Client) runHeartbeat(retryOpts retry.Options) {
+	var err error
 	for {
+		for r := retry.Start(retryOpts); r.Next(); {
+			// reconnect unless the error is nil or definitely temporary
+			if netErr, ok := err.(net.Error); err != nil && !(ok && netErr.Temporary()) {
```

Comment:
I would be surprised if we got a temporary error here, since this error is for the heartbeat rpc rather than connect. I don't think you can get a temporary error once a connection is established unless you call `SetDeadline`.